This is a validation pipeline for checking the accuracy of an assembler. 
The inputs to the pipeline are fasta files. A fasta file for the reference genome and one or more fasta files for contigs/reads.

This pipeline does the following things - 

1. Uses simple exact matching to get the alignment percent of a read/contig with the reference
2. Uses BWA to find the alignment of read/contig with the reference
3. Compares the percentage of kmers common between the reads/contig and reference

##PREREQUISITES -

Python 2.7


BWA - version used bwa-0.7.15-r1140, Should be present in the PATH

	Install using homebrew - brew install bwa

DSK - version used 2.1.0, Should be present in the PATH

	Install using homebrew - brew install dsk

SAMTOOLS - version used 1.3.1, Should be present in the PATH

	Install using homebrew - brew install samtools

##INSTALLATION

To use the program

SSH - 
	git clone git@github.com:mayankpahadia1993/validationpipeline.git


HTTPS - 
	git clone https://github.com/mayankpahadia1993/validationpipeline.git

##TESTING
			
	cd validationpipeline/test/;
	bash test.sh;

It should print "EVERYTHING IS WELL" at the end. Or otherwise it will give an error.


##USAGE 


	bash run.sh -r referenceFile.fa -s suffixTreeOutput.p -i InputFile.fa -a AlignmentResults

##OPTIONS

-r -s -i|-f are compulsory options. -i can be multiple


`-r | --reference filename` for passing the reference file


`-s | --suffixtree filename` filename will store the suffixtree


`-i | --input filename` filename contains the input file to be checked


`-a | --alignment filename` filename is the file in which alignment results will be shown. Default Value is alignmentResults.txt


`-f | --file filename` filename is a file which contains the input filenames. One filename per line.


`-suffixskip BinaryInteger` to skip the Suffix Tree Creation. The default value is 0 and will create the Suffix Tree


`-bwaskip BinaryInteger` to skip the bwa step. The default value is 0 and it will do the bwa step


`-kmer-size Integer` The integer given here would be chosen as the kmer size. Default is 30. 


`-abundance-min Integer` The integer given here would be chosen as the abundance min. Default is 3. Abundance min represents the minimum count of a kmer which will be stored. 


##INPUT

The mandatory input to the program is the reference file, suffix tree file and the files containing reads and contigs.

The reference file and read files must be in the fasta format. The reads here usually mean contigs or unitigs which we get after running an assembler.

Suffix Tree file is the filename in which the suffix tree gets stored, if it has been created in the pipeline. The creation works if the '-suffixskip 0' option is set, which is also the default option. 

If '-suffixskip 1' is used, the suffix tree creation step will be skipped and the suffix tree in the file mentioned with the option -s|--suffixtree will be used. 

##OUTPUT

These are the output files that are generated by the pipeline - 

1. AlignmentResults
2. *.commonKmers12
3. *.commonKmers12
4. SuffixTreeOutput file


AlignmentResults -  The output is in the file specified with the option "-a|--alignment". The file will contain a maximum of 5 columns all separated by tabs.

Example - 


FILENAME	Align Percentage	BWA Percentage	Percentage of Kmers from Reference File in Input File	Percentage of Kmers from Input File in Reference File

test/reads.fa	22.136	100.00%	100.0%	1.27982540829%

The output has 5 columns - 

	a. Filename - This column has the name of the input files(reads/contigs).


	b. Align Percentage - This column has the align percentage calculated using simple exact matching.


	c. BWA Percentage - This column has the align percentage calculated using BWA MEM.


	d. Percentage of Kmers from Reference File in Input File - This column has the percentage of kmers common from the reference file in the input file.


	e. Percentage of Kmers from Input File in Reference File - This column has the percentage of kmers common from the input file in the reference file.


*.commonKmers12 - These files are generated for all the input files. In this file, there is one row for one kmer from the input file. Every row contains two columns. The first column is the kmer and the second column is a binary number in which '1' represents the kmer is present in the reference file and '0' represents it is not. It is tab separated. 


*.commonKmers21 - These files are generated for all the input files. In this file, there is one row for one kmer from the reference file. Every row contains two columns. The first column is the kmer and the second column is a binary number in which '1' represents the kmer is present in the input file and '0' represents it is not. It is tab separated.

suffixTreeOutput File - This file stores the suffix tree of the reference which was created. This file is generated only if suffixskip option is at 0 i.e. '-suffixskip 0' option is set.


##Acknowledgement

dsk - https://github.com/GATB/dsk

Suffix tree - https://github.com/kvh/Python-Suffix-Tree

bwa - https://github.com/lh3/bwa

samtools - https://github.com/samtools
